### PR TITLE
Reduce memory usage of match all bitset

### PR DESCRIPTION
**Pull request by dnhatn**
By default, Elasticsearch uses up to 1 bit per document to store the set of root documents when the index mapping has nested fields. This PR introduces a special BitSet using less memory for match_all filters. This optimization is only triggered when the index mapping has a nested field, but that field never exists in documents.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/lucene/util/MatchAllBitSet.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        bits = new FixedBitSet(numBits);
+        bits.set(0, bits.length());
+    }
+
+    @Override
+    public void clear(int i) {
+        if (bits == null) {
+            initializeBitSet();
+        }
+        bits.clear(i);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/lucene/util/MatchAllBitSet.java:57]: Do we actually need to support clearing bits on this `MatchAllBitSet`? It's counter intuitive to me that something called `MatchAllBitSet` could have some of its bits not set.

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/lucene/util/MatchAllBitSet.java]: Good call. I've pushed https://github.com/elastic/elasticsearch/commit/0f12598c31a46579677a277341c08c2c98cbc3da to make it read-only.

**Review by jpountz — APPROVED** [review]
This looks fine to me. For the record, I'm in general not a fan of implementing optimizations for cases that should be as rare as this one, but this concern is alleviated by the fact that it shares logic with the same optimization for DLS where I would expect this condition to be less rare.

**Comment by dnhatn** [comment]
Thanks, Adrien. I was also hesitant to open this PR, but this class already exists in the security module.

**Pull request merged**